lib/Parser.js

```diff
@@ -1396,6 +1388,214 @@ class Parser extends Tapable {
 		}
 	}
 
+	_prewalkNode(node) {
```

Comment:
Instead of one `(pre)walkNode` method, which can handle all node types, you could move the switch into `(pre)walkStatement` `walkExpression` and `walkPattern` and only test for types that can occur there. `...Statement` `...Declaration` -> `(pre)walkStatement` , `...Expression` -> `walkExpression`, `...Pattern` -> `walkPattern`.

Reply:
I'm having a hard time doing this... Since previous methods were highly dynamic, I'm afraid of breaking anything...

Reply:
Where should I move `RestElement` and `SpreadElement`?

Reply:
I moved them `Expression`